api-only/pom.xml

```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2019 Oracle and/or its affiliates. All rights reserved.
+
+    This program and the accompanying materials are made available under the
+    terms of the Eclipse Distribution License v. 1.0, which is available at
+    http://www.eclipse.org/org/documents/edl-v10.php.
+
+    SPDX-License-Identifier: BSD-3-Clause
+
+-->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.glassfish.gmbal</groupId>
+        <artifactId>gmbal-project</artifactId>
+        <version>4.0.1-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <groupId>org.glassfish.gmbal</groupId>
+    <artifactId>gmbal-api-only</artifactId>
+    <version>4.0.1-SNAPSHOT</version>
+
+    <name>GMBAL (API only)</name>
+    <description>GlassFish MBean Annotation Library (API only)</description>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.glassfish.external</groupId>
+            <artifactId>management-api</artifactId>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.groupId}.${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.name}</Bundle-Name>
+                        <Bundle-Version>${project.version}</Bundle-Version>
+                        <Export-Package>org.glassfish.gmbal</Export-Package>
+                    </instructions>
+                </configuration>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>manifest</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            
+            <plugin> 
```

Comment:
Do you really mean to build the sources jar on every build?

Reply:
yes because impl jar depends on these sources - it unzips them to temporary folder so they can be recompiled with appropriate classpath

Reply:
So the impl jar doesn't simply reference the api-only jar? Wouldn't it be cleaner to use the shade or assembly plugin, rather than recompiling the sources?

Reply:
I agree that simple reference to api-only jar without the need of shade or assembly plugin would be cleaner solution. Unfortunately the class dependency tree of the ManagedObjectManager goes through ObjectRegistrationManager->Named->TimerFactory to some ~30+ other files from pfl-tf. While I could add [this particular method](https://github.com/eclipse-ee4j/orb-gmbal-pfl/blob/master/pfl-tf/src/main/java/org/glassfish/pfl/tf/timer/spi/TimerEventControllerBase.java#L37) to the interface and replace [the class argument](https://github.com/eclipse-ee4j/orb-gmbal-pfl/blob/master/pfl-tf/src/main/java/org/glassfish/pfl/tf/timer/spi/TimerFactory.java#L93) with an interface to remove the dependency on the `impl` package, to lower the class tree to ~16 classes, it is still too much, IMO.

I personally hate using the shade plugin - while it provides some useful functionality, 95% of it can be done by assembly plugin; another drawback of this plugin is that it may transform runtime project object model in an unpredictable way (ie mvn help:effective-pom may become useless)

Assembly is good candidate but generating OSGi headers becomes complicated (unless they're written by hand).

I personally prefer recompilation, which is not that big overhead in this particular project, just to make sure everything is correct. Alternative I can come up with is to just repackage required class files instead of doing full recompilation, WDYT?

Btw: in an ideal world, there would be just a dependency from impl to api jar and no repackaging (aka split package problems on JDK9+) but this would very likely require (possibly non-trivial) changes in downstream projects

Reply:
I agree that this is a bit of a mess.

If the additional classes for the impl are in the same package, the dependency idea will indeed cause problems. 

OK, so this is a legacy issue, and I don't think repackaging (which isn't standard in Maven anyway) is a real improvement, so let's go with your approach.